File: sdks/go/canned_replies.go
package suppuo

import (
	"context"
	"net/url"
)

// CannedRepliesResource manages per-workspace saved reply snippets
// (Bearer auth).
type CannedRepliesResource struct {
	c *Client
}

// List calls GET /api/v1/canned-replies.
func (r *CannedRepliesResource) List(ctx context.Context) ([]CannedReply, error) {
	var out struct {
		CannedReplies []CannedReply `json:"cannedReplies"`
	}
	if err := r.c.do(ctx, "GET", "/api/v1/canned-replies", nil, nil, false, &out); err != nil {
		return nil, err
	}
	return out.CannedReplies, nil
}

// CannedReplyInput is the payload for CannedReplies.Create.
type CannedReplyInput struct {
	Title string `json:"title"`
	Body  string `json:"body"`
}

// Create calls POST /api/v1/canned-replies.
func (r *CannedRepliesResource) Create(ctx context.Context, input CannedReplyInput) (*CannedReply, error) {
	var out CannedReply
	if err := r.c.do(ctx, "POST", "/api/v1/canned-replies", nil, input, false, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// CannedReplyPatch is the payload for CannedReplies.Update. Nil fields
// are omitted.
type CannedReplyPatch struct {
	Title *string `json:"title,omitempty"`
	Body  *string `json:"body,omitempty"`
}

// Update calls PATCH /api/v1/canned-replies/:id.
func (r *CannedRepliesResource) Update(ctx context.Context, id string, patch CannedReplyPatch) (*CannedReply, error) {
	var out CannedReply
	path := "/api/v1/canned-replies/" + url.PathEscape(id)
	if err := r.c.do(ctx, "PATCH", path, nil, patch, false, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Delete calls DELETE /api/v1/canned-replies/:id.
func (r *CannedRepliesResource) Delete(ctx context.Context, id string) error {
	path := "/api/v1/canned-replies/" + url.PathEscape(id)
	return r.c.do(ctx, "DELETE", path, nil, nil, false, nil)
}

File: backend/src/routes/canned-replies.ts
import { Router } from 'express';
import { z } from 'zod';
import { prisma } from '../lib/db.js';
import { newId } from '../lib/ids.js';
import { sendOk, sendCreated, sendErr } from '../lib/http.js';
import { h as asyncHandler } from '../lib/async-handler.js';

// /api/v1/canned-replies — per-workspace saved reply snippets.

const router = Router();

const upsertBody = z.object({
  title: z.string().trim().min(1).max(120),
  body: z.string().trim().min(1).max(20_000),
});

router.get(
  '/',
  asyncHandler(async (req, res) => {
    const rows = await prisma.cannedReply.findMany({
      where: { accountId: req.auth!.accountId as string },
      orderBy: { title: 'asc' },
    });
    sendOk(res, req, { cannedReplies: rows });
  }),
);

router.post(
  '/',
  asyncHandler(async (req, res) => {
    const input = upsertBody.parse(req.body);
    const row = await prisma.cannedReply.create({
      data: {
        id: newId('cnr'),
        accountId: req.auth!.accountId as string,
        title: input.title,
        body: input.body,
      },
    });
    sendCreated(res, req, row);
  }),
);

router.patch(
  '/:id',
  asyncHandler(async (req, res) => {
    const input = upsertBody.partial().parse(req.body);
    const accountId = req.auth!.accountId as string;
    const existing = await prisma.cannedReply.findFirst({
      where: { id: String(req.params.id), accountId },
    });
    if (!existing) return sendErr(res, req, 404, 'NOT_FOUND', 'canned reply not found');
    const row = await prisma.cannedReply.update({
      where: { id: existing.id },
      data: input,
    });
    sendOk(res, req, row);
  }),
);

router.delete(
  '/:id',
  asyncHandler(async (req, res) => {
    const accountId = req.auth!.accountId as string;
    const existing = await prisma.cannedReply.findFirst({
      where: { id: String(req.params.id), accountId },
    });
    if (!existing) return sendErr(res, req, 404, 'NOT_FOUND', 'canned reply not found');
    await prisma.cannedReply.delete({ where: { id: existing.id } });
    sendOk(res, req, { deleted: true });
  }),
);

export default router;
